### PR TITLE
Add `requires_proxy_auth` arg to webhooks

### PR DESCRIPTION
**Pull request by 2022tgoel**
## Describe your changes

- _Provide Linear issue reference (e.g. MOD-1234) if available._

<details> <summary>Backward/forward compatibility checks</summary>

---

Check these boxes or delete any item (or this section) if not relevant for this PR.

- [ ] Client+Server: this change is compatible with old servers
- [ ] Client forward compatibility: this change ensures client can accept data intended for later versions of itself

Note on protobuf: protobuf message changes in one place may have impact to
multiple entities (client, server, worker, database). See points above.

---

</details>

### CODE REVIEW

**Review by luiscape — CHANGES_REQUESTED** [review]
We are missing a few places: `_wsgi_app`, `_web_server`, etc. Can you find instances of `api_pb2.WebhookConfig` and add this parameter everywhere?

File: modal/partial_function.py [hunk 1]
```diff
@@ -332,6 +334,7 @@ def _asgi_app(
     *,
     label: Optional[str] = None,  # Label for created endpoint. Final subdomain will be <workspace>--<label>.modal.run.
     custom_domains: Optional[Iterable[str]] = None,  # Deploy this endpoint on a custom domain.
+    requires_proxy_auth: bool = False,  # Require Proxy-Authenticate HTTP Headers on requests to the endpoint
```
Comment (luiscape) [inline modal/partial_function.py:337]: `Proxy-Authenticate` is a response header. This requires the `Proxy-Authorization` request header.

**Review by luiscape — APPROVED** [review]
Please update comments as per review. Good to merge after. 

@mwaskom is this alright with you?

File: modal/partial_function.py (hunk 1 above)
Comment (2022tgoel) [reply modal/partial_function.py]: 👍

File: modal/partial_function.py [hunk 2]
```diff
@@ -332,6 +334,7 @@ def _asgi_app(
     *,
     label: Optional[str] = None,  # Label for created endpoint. Final subdomain will be <workspace>--<label>.modal.run.
     custom_domains: Optional[Iterable[str]] = None,  # Deploy this endpoint on a custom domain.
+    requires_proxy_auth: bool = False,  # Require Proxy-Authorization HTTP Headers on requests to the endpoint
```
Comment (mwaskom) [inline modal/partial_function.py:337]: Nit, but is "the endpoint" ambiguous for the asgi_app/wsgi_app/web_server decorators?

**Review by mwaskom — COMMENTED** [review]
Couple questions:

- Would it make sense to cover this in the unit tests for these decorators?
- Do we need to do this in two steps (i.e., a proto PR, then do the client interface PR after we've enabled server support?)
- Can we add a changelog entry?

**Comment by mwaskom** [comment]
Oops ... guess I'm a little later there.

**Pull request merged**